cirq-core/cirq/ops/dense_pauli_string.py

```diff
@@ -205,67 +200,69 @@ def __iter__(self) -> Iterator['cirq.Gate']:
         for i in range(len(self)):
             yield self[i]
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.pauli_mask)
 
-    def __neg__(self):
-        return DensePauliString(coefficient=-self.coefficient, pauli_mask=self.pauli_mask)
+    def __neg__(self: TCls) -> TCls:
+        return type(self)(coefficient=-self.coefficient, pauli_mask=self.pauli_mask)
 
-    def __truediv__(self, other):
-        if isinstance(other, (sympy.Basic, numbers.Number)):
+    def __truediv__(self: TCls, other: Union[sympy.Basic, int, float, complex]) -> TCls:
+        if isinstance(other, (sympy.Basic, int, float, complex)):
             return self.__mul__(1 / other)
 
         return NotImplemented
 
-    def __mul__(self, other):
-        if isinstance(other, BaseDensePauliString):
-            max_len = max(len(self.pauli_mask), len(other.pauli_mask))
-            min_len = min(len(self.pauli_mask), len(other.pauli_mask))
+    def __mul__(self: TCls, other: Union[sympy.Basic, int, float, complex, TCls]) -> TCls:
+        cls = type(self)
+
+        if isinstance(other, cls):
```

Comment:
Does this mean multiplying a `MutableDensePauliString` instance with a `DensePauliString` was allowed before this change but will not be allowed anymore? This sounds like a breaking change and should not be done. 

For binary operations that expect an `other` argument which can be a base dense pauli string, I think we should default to returning a `MutableDensePauliString` if the types of self and other are not same (but both are instances of BaseDensePauliString), else return `type(self)`.  


This should be done for `__mul__`,  `__rmul__`, `tensor_product` etc. Please also add corresponding tests.

Reply:
I added the possibility to cross operate on mutable and non-mutable dense Pauli strings.